### DISCUSSION TITLE
Use with Unity3D

### DISCUSSION THREAD

**Discussion by lorenblue** [ArxOne/MrAdvice#194] [General]
Hi there!

I am just now learning about weaving and aspects and such, so forgive me. Is there any way I can get this to work with Unity (the game engine)? I can add the nuget package just fine, and have no compiler errors, but when I try to copy the sample aspect and call a decorated method, the advise method never fires. 

Cheers,

Lorenzo

**Comment by picrap** [comment]
Hi Lorenzo,

it should work (but I didn’t test 😁). Could you provide a build log (in normal verbosity)?

Pascal

**Comment by lorenblue** [comment]
Ok so the difficulty here (aside from my complete lack of understanding on this sort of thing) is that I believe Unity uses its own proprietary compiler, and I would imagine that the build process is not a conventional one (at least from my perspective, as it doesn't at all involve me manually instructing my IDE to build the project/solution). The best I could find so far is the Unity Editor log, which I captured after Unity compiled my assembly:

Asset Pipeline Refresh: Total: 0.018 seconds - Initiated by RefreshV2(AllowForceSynchronousImport)
Asset Pipeline Refresh: Total: 0.019 seconds - Initiated by RefreshV2(NoUpdateAssetOptions)
Registering precompiled user dll's ...
Registered in 0.008419 seconds.
Start importing Assets/_Project/Scripts/TESTATTRIBUTE.cs using Guid(66e9f0996a3f4784a3bded3be8eb9051) Importer(-1,00000000000000000000000000000000)  -> (artifact id: 'f7b6819319510a42db64b817d150cf77') in 0.010537 seconds 
[ScriptCompilation] Requested script compilation because: Assetdatabase observed changes in script compilation related files
AssetDatabase: script compilation time: 0.083088s
Refreshing native plugins compatible for Editor in 0.91 ms, found 12 plugins.
Preloading 1 native plugins for Editor in 0.09 ms.
Asset Pipeline Refresh: Total: 0.209 seconds - Initiated by RefreshV2(NoUpdateAssetOptions)
Starting: C:\Program Files\Unity\Hub\Editor\2021.3.16f1\Editor\Data\bee_backend.exe --profile="Library/Bee/backend_profiler0.traceevents" --stdin-canary --dagfile="Library/Bee/1300b0aE.dag" --continue-on-failure ScriptAssemblies
WorkingDir: C:/Users/Lorenzo/Documents/Unity Projects/Vision Training
ExitCode: 0 Duration: 0s913ms
[778/790    0s] Csc Library/Bee/artifacts/1300b0aE.dag/VisionTraining.dll (+2 others)
[788/790    0s] CopyFiles Library/ScriptAssemblies/VisionTraining.pdb
[789/790    0s] CopyFiles Library/ScriptAssemblies/VisionTraining.dll
*** Tundra build success (0.84 seconds), 3 items updated, 790 evaluated
Reloading assemblies after finishing script compilation.
Registering precompiled user dll's ...
Registered in 0.007996 seconds.
Reloading assemblies after forced synchronous recompile.
Begin MonoManager ReloadAssembly
Symbol file LoadedFromMemory doesn't match image C:\Users\Lorenzo\Documents\Unity Projects\Vision Training\Library\PackageCache\com.unity.visualscripting@1.7.8\Editor\VisualScripting.Core\Dependencies\DotNetZip\Unity.VisualScripting.IonicZip.dll
Symbol file LoadedFromMemory doesn't match image C:\Users\Lorenzo\Documents\Unity Projects\Vision Training\Library\PackageCache\com.unity.visualscripting@1.7.8\Editor\VisualScripting.Core\Dependencies\YamlDotNet\Unity.VisualScripting.YamlDotNet.dll
Native extension for WindowsStandalone target not found
Native extension for Android target not found
Refreshing native plugins compatible for Editor in 0.98 ms, found 12 plugins.
Preloading 1 native plugins for Editor in 0.09 ms.
[MODES] ModeService[none].Initialize
[MODES] ModeService[none].LoadModes
[MODES] Loading mode Default (0) for mode-current-id-Vision Training
Mono: successfully reloaded assembly
- Completed reload, in  2.350 seconds
Domain Reload Profiling:
	ReloadAssembly (2351ms)
		BeginReloadAssembly (410ms)
			ExecutionOrderSort (0ms)
			DisableScriptedObjects (19ms)
			BackupInstance (0ms)
			ReleaseScriptingObjects (0ms)
			CreateAndSetChildDomain (222ms)
		EndReloadAssembly (1848ms)
			LoadAssemblies (97ms)
			RebuildTransferFunctionScriptingTraits (0ms)
			SetupTypeCache (268ms)
			ReleaseScriptCaches (2ms)
			RebuildScriptCaches (69ms)
			SetupLoadedEditorAssemblies (895ms)
				LogAssemblyErrors (0ms)
				InitializePlatformSupportModulesInManaged (15ms)
				SetLoadedEditorAssemblies (0ms)
				RefreshPlugins (1ms)
				BeforeProcessingInitializeOnLoad (252ms)
				ProcessInitializeOnLoadAttributes (589ms)
				ProcessInitializeOnLoadMethodAttributes (33ms)
				AfterProcessingInitializeOnLoad (5ms)
				EditorAssembliesLoaded (0ms)
			ExecutionOrderSort2 (0ms)
			AwakeInstancesAfterBackupRestoration (300ms)
Platform modules already initialized, skipping
Registering precompiled user dll's ...
Registered in 0.008011 seconds.
Asset Pipeline Refresh: Total: 2.533 seconds - Initiated by StopAssetImportingV2(ForceSynchronousImport | ForceDomainReload)
	Summary:
		Imports: total=0 (actual=0, local cache=0, cache server=0)
		Asset DB Process Time: managed=0 ms, native=140 ms
		Asset DB Callback time: managed=30 ms, native=0 ms
		Scripting: domain reloads=1, domain reload time=2359 ms, compile time=1 ms, other=0 ms
		Project Asset Count: scripts=6629, non-scripts=2147
		Asset File Changes: new=0, changed=0, moved=0, deleted=0
		Scan Filter Count: 0
	InvokeBeforeRefreshCallbacks: 0.621ms
	ApplyChangesToAssetFolders: 0.060ms
	Scan: 0.000ms
	OnSourceAssetsModified: 0.001ms
	InitializeImportedAssetsSnapshot: 10.600ms
	GetAllGuidsForCategorization: 0.728ms
	CategorizeAssets: 95.681ms
	ImportOutOfDateAssets: 2368.382ms (2364.585ms without children)
		CompileScripts: 1.319ms
		EnsureUptoDateAssetsAreRegisteredWithGuidPM: 1.257ms
		InitializingProgressBar: 0.000ms
		PostProcessAllAssetNotificationsAddChangedAssets: 0.747ms
		OnDemandSchedulerStart: 0.475ms
	ReloadSourceAssets: 1.494ms
	UnloadImportedAssets: 3.198ms
	PostProcessAllAssets: 30.835ms
	GatherAllCurrentPrimaryArtifactRevisions: 0.292ms
	UnloadStreamsBegin: 0.236ms
	LoadedImportedAssetsSnapshotReleaseGCHandles: 1.589ms
	GetLoadedSourceAssetsSnapshot: 1.265ms
	PersistCurrentRevisions: 0.299ms
	UnloadStreamsEnd: 0.031ms
	GenerateScriptTypeHashes: 2.711ms
	Untracked: 48.379ms
Asset Pipeline Refresh: Total: 0.016 seconds - Initiated by RefreshV2(NoUpdateAssetOptions)
Asset Pipeline Refresh: Total: 0.019 seconds - Initiated by RefreshV2(AllowForceSynchronousImport)

**Comment by picrap** [comment]
Right, if that is a non-standard build, things will be complicated.
I suggest you take a look at https://github.com/ArxOne/MrAdvice/issues/152 and https://github.com/ArxOne/MrAdvice/issues/174
It seems that you will have to add a post-build step and invoke MrAdvice there.

**Comment by lorenblue** [comment]
Ah I see. When you say post-build step, is it good enough if I invoke MrAdvice after Unity notifies me that scripts have recompiled, or does it need to be at a specific (earlier) time?

**Comment by picrap** [comment]
Any time after the assembly is generated is good enough.

**Comment by lorenblue** [comment]
Nice! I do have nuget hooked up with Unity, but I assume the best (only) way would be to call it through the command line like in your docs. If this is the case, which parameters must I include? Assembly path seems necessary, but I'm afraid I don't really know what the others refer to, and thus if they're necessary or not.

**Comment by lorenblue** [comment]
Getting close.. Again just not sure of which other parameters to include and what exactly they refer to, sorry!

///////net461
22:53:25 ❯ .\MrAdvice.Weaver.exe AssemblyPath="C:\Users\Lorenzo\Documents\Unity Projects\Vision Training\Library\ScriptAssemblies\VisionTraining.dll"
! Can't load 'netstandard, Version=2.1.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
! Can't load 'UnityEngine.CoreModule, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
! Can't load 'MrAdvice, Version=2.0.0.0, Culture=neutral, PublicKeyToken=c0e7e6eab6f293d8'
! Can't load 'NaughtyAttributes.Core, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
! Can't load 'UnityEngine.UI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
! Can't load 'Unity.XR.Interaction.Toolkit, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
! Can't load 'DOTween, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
! IAdvice interface not found here (not referenced means not used), exiting

**Comment by lorenblue** [comment]
Nevermind, I was just missing the referencepath parameter it seems! Magnificent. Will report back if I run into any further issues :)

**Reply by picrap** [reply]
You are going too fast :). I was looking at parameters to give you a correct sample.
If you know any way to automate integration of MrAdvice with Unity, let me know (it’s a good feature to add).

**Comment by lorenblue** [comment]
Ok, so I can confirm that it does indeed work.. to an extent :P

It appears that invoking MrAdvice whenever I want is _not_ sufficient to apply the changes. Rather, I need to invoke MrAdvice and then force a recompile of scripts. I have tried all sorts of Unity callbacks to get my desired behaviour of hooking into the compilation process at the right time so I don't have to manually recompile, but to no avail so far. As a result, the only process that works is: 

Make changes to code in the assembly -> Wait for Unity to recompile -> Invoke MrAdvice -> Make Unity recompile again

As a secondary issue, I was hoping that I could use a method advice aspect to debug a message after an IEnumerator method (coroutine) has finished, but I understand that there may be some complexities there due to how Unity handles such methods. A message before execution worked no problem.

Finally, I was hoping to apply a method advice aspect to an abstract parent method and have it propagated to child classes, but this did not seem to work.

Phew, that's all for now!

**Comment by lorenblue** [comment]
Hey again! I'm wondering if you have had/do have any time to explore this further. As explained above, I had gotten really close to having it work before, save for those few lingering issues. I wish I could dive into it more, but I just don't have the time (nor brainpower). Thanks!

**Comment by picrap** [comment]
I don’t work with Unity3D, so it’s hard for me to implement it.
Regarding the `IEnumerable`, You probably need to wrap all enumeration.
